Fix #102: Define non-constrainable property

anssiko opened the pull request:
- Define non-constrainable property
- Apply the newly minted definition to:
  - focalLength
  - horizontalFieldOfView
  - verticalFieldOfView

anssiko committed:
Fix #102: Define non-constrainable property

Apply the newly minted definition to:

* focalLength
* horizontalFieldOfView
* verticalFieldOfView

anssiko commented:
Fixes #102.

jan-ivar commented on index.html:
```diff
@@ -560,15 +564,18 @@
             The <dfn><code>focalLength</code></dfn> dictionary member
-            represents the <a>depth map</a>'s <a>focal length</a>.
+            represents the <a>depth map</a>'s <a>focal length</a>. It is a
+            <a>non-constrainable property</a>.
```
`non-constrainable` seems like a misnomer. Constraints are properties of the _application_, as in "here are my constraints", a request. So it is perfectly reasonable to set constraints on inherent unchangeable properties. Remember `getUserMedia` may pick from several cameras/mics.

anssiko replied:
I copied the name from #102 :)

Would rename of the concept into "unchangeable property" make things better? I'm happy to refactor the PR further given concrete suggestions how to convey the read-only-ness of these in the context of the depth stream tracks. Currently, those constraints do not exist for the regular MSTs, but could be if upstreamed to a future version of the mediacapture-main.

jan-ivar replied:
Having things that only show up in getSettings violates the Constrainable pattern, so unless we have a good reason, I think we should mirror facingMode. Is there a downside?

jan-ivar replied:
But yeah you're right, I'm objecting to the concept, not the name.

anssiko replied:
I'd be happy to mirror it, if educated how to do so the right way. There were no hooks to related concepts that could be referenced, also the facingMode prose seemed not reusable, consider:

> This string (or each string, when a list) should be one of the members of VideoFacingModeEnum. The members describe the directions that the camera can face, as seen from the user's perspective. Note that getConstraints may not return exactly the same string for strings not in this enum. This preserves the possibility of using a future version of WebIDL enum for this property.

... and:

> A camera can report multiple facing modes. For example, in a high-end telepresence solution with several cameras facing the user, a camera to the left of the user can report both "left" and "user".

Concrete suggestions welcome!

jan-ivar replied:
Or just mirror any other constraint, like `frameRate` (a numerical one is probably closer to what you want). There's nothing special with facingMode in this regard. I.e. add to:
- MediaTrackSupportedConstraints
- MediaTrackCapabilities
- MediaTrackConstraintSet
- MediaTrackSettings

(These four places used to be a single dictionary, but people who are not me decided to split them up).

jan-ivar commented on index.html:
```diff
@@ -661,6 +668,33 @@
+          property</dfn> is ignored in the context of the
+          <a>applyConstraints()</a> algorithm. Its value can be obtained via
+          the <a>getSettings()</a> accessor. A <a>non-constrainable
+          property</a>'s <a>fitness distance</a> is always 0.
```
This new thing is not in the constrainable algorithm today, and seems unnecessary. I see nothing harmful in people putting constraints on fixed values. Worst case `applyConstraints` just fails.

anssiko replied:
Ok, so will drop the fitness distance special casing.

anssiko committed:
Revert "Fix #102: Define non-constrainable property"

This reverts commit 0ad89ba625d00cdf96371e0cb65e4698a951bed6.

anssiko committed:
Fix #102: Define constrainable properties (redux)

Properly define the constrainable properties:

* depthNear
* depthFar
* focalLength
* horizontalFieldOfView
* verticalFieldOfView

Reorganize subsections into a more logical order.

anssiko commented:
@jan-ivar Thanks for the good suggestions! I reverted 1c8f97d and did another attempt to define the constrainable properties per your suggestion. Please review 66fab9e.

anssiko committed:
No special value (0) for constrainable properties

anssiko committed:
Use (double or DoubleRange) for MediaTrackCapabilities

anssiko commented:
@jan-ivar Thanks for the feedback!
- The unit discussion happened in #113. Meter seem like a more natural unit for the near and far values, while millimeter is the de facto unit for the focal length. I think this is a reasonable tradeoff to make. In the common use cases you'd use near and far but not focal length. In advanced use cases, the web developers know how to convert between the units.
- The names were initially `near` and `far`. We prefixed them with `depth` based on feedback, see #109.

Changes:
- Removed the use of a special value (`0`) for constrainable properties in 3bb0bbe
- Used `(double or DoubleRange)` for `MediaTrackCapabilities` in 31ef274

Unless other comments arise, I'd like to land this PR.

jan-ivar commented:
Sorry I don't see any feedback in https://github.com/w3c/mediacapture-depth/pull/109. All I see is @dontcallmedom saying in https://github.com/w3c/mediacapture-depth/issues/103 that `format` is too generic, with which I agree. I don't think that criticism extends to `near` and `far` though or he would have mentioned it I think. I don't think he meant to blanket-prefix constraints with their domain, as we've never done that before. I can't think of another domain for `near` and `far` either (the only other domain I can think of would be time, but something like `timeNear` and `timeFar` would be a poor metaphor of distance applied time compared to better words like 'soon' and `late`), so I think we would be fine with `near` and `far` FWIW. Otherwise LGTM. @dontcallmedom thoughts?

anssiko committed:
Drop depth prefix from near and far constraints

anssiko commented:
@jan-ivar I dropped the depth prefix per your suggestion in b0b81df. Thanks for the review, landing now!
